src/horust/supervisor/process_spawner.rs

```diff
@@ -172,14 +175,57 @@ fn redirect_output(
 
 /// Find program on PATH.
 ///
-fn find_program_path(program_name: &String) -> std::result::Result<String, Errno> {
-    for dir in std::env::var("PATH").unwrap().split(":") {
-        let path = format!("{}/{}", dir, program_name);
-        if std::path::Path::new(&path).exists() {
-            return Ok(path);
+#[derive(Debug)]
+enum PathSearchError {
+    NotFound,
+    InvalidPath,
+    MissingPath,
+}
+
+impl std::error::Error for PathSearchError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        // Implement the source method to return the underlying error, if any.
+        None
+    }
+}
+
+impl std::fmt::Display for PathSearchError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PathSearchError::NotFound => {
+                write!(f, "Program not found in any of the PATH directories")
+            }
+            PathSearchError::InvalidPath => write!(f, "Invalid path for program"),
+            PathSearchError::MissingPath => write!(f, "PATH environment variable is not set"),
         }
     }
-    Err(ENOENT)
+}
+
+fn find_program(program_name: &String) -> Result<String, PathSearchError> {
+    // Get the PATH environment variable
+    let path_var = match std::env::var_os("PATH") {
+        Some(val) => val,
+        None => return Err(PathSearchError::MissingPath),
+    };
+
+    // Split the PATH variable into individual paths
+    let paths: Vec<PathBuf> = std::env::split_paths(&path_var).collect();
+
+    // Iterate through each directory in PATH and check if the program exists
+    for path in paths {
+        let program_path = path.join(program_name);
+
+        // Check if the program file exists at this path
+        if program_path.is_file() {
+            if let Some(path_str) = program_path.to_str() {
+                return Ok(path_str.to_string());
+            } else {
+                return Err(PathSearchError::InvalidPath);
```

Comment:
how can this be triggered? I think you're trying to map an Option to a Result, for that you can use return program_path.to_str().ok_or_else(||{anyhow!() });

Reply:
This can only be triggered if there is an encoding error in the PATH or the program name, and I think it is fine to panic as the problem must be fixed before Horust can run.